server/datastore/s3/installer.go

```diff
@@ -0,0 +1,62 @@
+package s3
+
+import (
+	"context"
+	"path"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/fleetdm/fleet/v4/server/config"
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+)
+
+// InstallerParams describes the params used to locate an installer in an
+// S3 bucket
+type InstallerParams struct {
+	secret  string
+	ext     string
+	desktop bool
+}
+
+func (p InstallerParams) buildKey() string {
+	var dir string
+	if p.desktop {
+		dir = "desktop"
+	}
+	return path.Join(p.secret, dir, "fleet-osquery."+p.ext)
+}
+
+// InstallerStore contains methods to retrieve installers from S3
+type InstallerStore struct {
+	*datastore
+}
+
+// NewInstallerStore creates a new instance with the given S3 config
+func NewInstallerStore(config config.S3Config) (*InstallerStore, error) {
+	s3Store, err := newDatastore(config)
+	if err != nil {
+		return nil, err
+	}
+
+	return &InstallerStore{s3Store}, nil
+}
+
+// Exists checks if an installer exists in the S3 bucket
+func (s *InstallerStore) Exists(ctx context.Context, params InstallerParams) bool {
+	key := params.buildKey()
+	_, err := s.s3client.HeadObject(&s3.HeadObjectInput{Bucket: &s.bucket, Key: &key})
+	return err != nil
+}
+
+// GetLink returns a pre-signed S3 link that can be used to download the
```

Comment:
Sorry, I forgot to more clearly define this:
Should we instead serve the package itself from Fleet? `client --> Fleet --> S3`, vs. this code that does `client --> S3`.

Let's check with @zwinnerman-fleetdm.
If we do it this way then S3 will need to be publicly accessible, etc.

Reply:
My understanding after reading [this](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-presigned-url.html) is that pre-signed URLs are accessible to whomever has it, so there shouldn't be extra work but worth also double checking with @zwinnerman-fleetdm / @rfairburn

Reply:
I went ahead and changed this function to return a blob with the file instead, we can optimize this with a pre-signed link later if we want to.